thrust/testing/catch2_test_transform.cu

```diff
@@ -0,0 +1,752 @@
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/iterator/retag.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/pair.h>
+#include <thrust/transform.h>
+#include <thrust/tuple.h>
+
+#include "catch2_test_helper.h"
+#include "unittest/random.h"
+#include "unittest/special_types.h"
+
+TEMPLATE_LIST_TEST_CASE("UnarySimple", "[transform]", vector_list)
+{
+  using Vector = TestType;
+  using T      = typename Vector::value_type;
+
+  typename Vector::iterator iter;
+
+  Vector input{1, -2, 3};
+  Vector output(3);
+  Vector result{-1, 2, -3};
+
+  iter = thrust::transform(input.begin(), input.end(), output.begin(), ::cuda::std::negate<T>());
+
+  CHECK(std::size_t(iter - output.begin()) == input.size());
+  CHECK(output == result);
+}
+
+template <typename InputIterator, typename OutputIterator, typename UnaryFunction>
+OutputIterator transform(my_system& system, InputIterator, InputIterator, OutputIterator result, UnaryFunction)
+{
+  system.validate_dispatch();
+  return result;
+}
+
+TEST_CASE("UnaryDispatchExplicit", "[transform]", )
+{
+  thrust::device_vector<int> vec(1);
+
+  my_system sys(0);
+  thrust::transform(sys, vec.begin(), vec.begin(), vec.begin(), 0);
+
+  CHECK(sys.is_valid());
+}
+
+template <typename InputIterator, typename OutputIterator, typename UnaryFunction>
+OutputIterator transform(my_tag, InputIterator, InputIterator, OutputIterator result, UnaryFunction)
+{
+  *result = 13;
+  return result;
+}
+
+TEST_CASE("UnaryDispatchImplicit", "[transform]", )
+{
+  thrust::device_vector<int> vec(1);
+
+  thrust::transform(
+    thrust::retag<my_tag>(vec.begin()), thrust::retag<my_tag>(vec.begin()), thrust::retag<my_tag>(vec.begin()), 0);
+
+  CHECK(13 == vec.front());
+}
+
+TEMPLATE_LIST_TEST_CASE("UnaryNoStencilSimple", "[transform_if]", vector_list)
+{
+  using Vector = TestType;
+  using T      = typename Vector::value_type;
+
+  typename Vector::iterator iter;
+
+  Vector input{0, -2, 0};
+  Vector output{-1, -2, -3};
+  Vector result{-1, 2, -3};
+
+  iter =
+    thrust::transform_if(input.begin(), input.end(), output.begin(), ::cuda::std::negate<T>(), ::cuda::std::identity{});
```

Comment:
Nitpick: I hate that we rely on implicit conversion from integer to bool here and also that 0 converts to false, especially given that there is no relation to negate.

Reply:
:)